algopy_testing/src/algopy_testing/context.py

```diff
@@ -168,54 +232,103 @@ def update_asset(self, asset_id: int, **asset_fields: Unpack[AssetFields]) -> No
+            app_id (int): Application ID.
 
         Returns:
-            algopy.Application: The application associated with the provided ID.
+            algopy.Application: The application associated with the ID.
         """
-        from algopy import Application, UInt64
+        import algopy
 
-        app_id = int(app_id) if isinstance(app_id, UInt64) else app_id
+        app_id = int(app_id) if isinstance(app_id, algopy.UInt64) else app_id
 
         if app_id not in self.application_data:
             raise ValueError("Application not found in testing context!")
 
-        return Application(app_id)
+        return algopy.Application(app_id)
 
     def update_application(
         self, app_id: int, **application_fields: Unpack[ApplicationFields]
     ) -> None:
         """
-        Update an existing application in the test context.
+        Update an existing application.
 
         Args:
-            app_id (int): The ID of the application to update.
-            application (algopy.Application): The new application data.
+            app_id (int): Application ID.
+            **application_fields: New application data.
         """
         if app_id not in self.application_data:
             raise ValueError("Application not found in testing context!")
 
         self.application_data[app_id].update(application_fields)
 
-    def add_inner_transaction(self, itxn: BaseInnerTransaction) -> None:
+    def add_contract(
+        self,
+        contract: algopy.Contract | algopy.ARC4Contract,
+    ) -> None:
         """
-        Add an inner transaction to the test context.
+        Add a contract to the context.
 
         Args:
-            itxn (BaseInnerTransaction): The inner transaction to be added.
+            contract (algopy.Contract | algopy.ARC4Contract): The contract to add.
+        """
+        self.contracts.append(contract)
+
+    def append_inner_transaction_group(
+        self,
+        itxn: typing.Sequence[
+            algopy.itxn.InnerTransactionResult
+            | algopy.itxn.PaymentInnerTransaction
+            | algopy.itxn.KeyRegistrationInnerTransaction
+            | algopy.itxn.AssetConfigInnerTransaction
+            | algopy.itxn.AssetTransferInnerTransaction
+            | algopy.itxn.AssetFreezeInnerTransaction
+            | algopy.itxn.ApplicationCallInnerTransaction
+        ],
+    ) -> None:
+        """
+        Append a group of inner transactions to the context.
+
+        Args:
+            itxn (Sequence[InnerTransactionResult | PaymentInnerTransaction |
+            KeyRegistrationInnerTransaction | AssetConfigInnerTransaction
+            | AssetTransferInnerTransaction | AssetFreezeInnerTransaction
+            | ApplicationCallInnerTransaction]): The group of inner transactions to append.
+        """
+        import algopy.itxn
+
+        self.inner_transaction_groups.append(cast(list[algopy.itxn.InnerTransactionResult], itxn))
```

Comment:
rather than casting, could make a union type at the module level `InnerTransactionResultType = algopy.itxn.InnerTransactionResult | algopy.itxn.PaymentTransactionResult | ...`

Then make the list on the context this type
selft.inner_transaction_groups `list[Sequence[InnerTransactionResultType]]`

and lastly make the itxn arg `Sequence[InnerTransactionResultType]`

Reply:
@daniel-makerx i ended up only addressing this partially, the reason i went for the cast in the first place is because the algopy.itxn.InnerTransactionResult is the only non protocol class in the itxn module. The rest are all non runtime-checkable protocols, hence using a tuple works well but the instance user receives won't work with methods like isinstance and you will have to do some sort of casting regardless if you get a union type in return. I still apply the suggestion on InnerTransactionResultType but only on the append method to make it more concise, the inner_transaction_group is using Sequence as suggested however i still do the casting and returning object type from this method is algopy.itxn.InnerTransactionResult, for type checking user can refer to the type enum and InnerTransactionResult conveniently encapsulates all fields from all transactions. 

Let me know what you think on this @daniel-makerx